### ISSUE TITLE
Erroneous urldecode_rfc3986

### ISSUE THREAD

**Issue by rapotek** [opencaching/okapi#535]
The description of `urldecode_rfc3986` in `okapi/okapi/core/OAuth/OAuthUtil.php` states that:

> // This decode function isn't taking into consideration the above
> // modifications to the encoding process. However, this method doesn't
> // seem to be used anywhere so leaving it as is.

But this method IS used. For example in `OAuthUtil::parse_parameters()` and then in `OAuthRequest::from_request()`. The missing modifications corresponding to `urlencode_rfc3986` causes f.ex. that oauth_signature passed in URL, which contain space encoded as '+' is treated as invalid because after decode there are spaces instead of '+' while signature built using `urlencode_rfc3986` converts spaces to '+'.

Therefore I think the method body should be as follows:
```
return str_replace(
    ' ',
    '+',
    str_replace('~', '%7E', rawurldecode($string))
);
```
but I suppose someone should confirm that I am not wrong.

**Comment by following5** [comment]
Have a close look at function `OAuthUtil::urlencode_rfc3986()`:

```
    return str_replace(
        '+',
        ' ',
        str_replace('%7E', '~', rawurlencode($input))
    );
```

It first does a ` rawurlencode()`, then _decodes_* the spaces and `~`.  So those two characteres effectively are not encoded. `urldecode_rfc3986()` then will decode the remaing encoded charactes (which acually are only `/` and `=`, because OAuth params effectively consist of base64 charset).

And `urldecode()`,  which is called from `urldecode_rfc3986()`, will passthrough the spaces and ~ and just decode the rest.

So the only thing wrong about `OAuthUtil::urldecode_rfc3986()` is the comment.

<raw>*</raw> Actually, `rawurlencode()` - differently from `urlencode()` - encodes spaces to %20 instead of +. So this + => space translation is unnecessary here. But it is failsafe, for the case someone would replace `rawurlencode()` by `urlencode()`.

**Comment by following5** [comment]
fixed by c9f4f8e
